morpheus/utils/directory_watcher.py

```diff
@@ -0,0 +1,214 @@
+
+        # Will be a watchdog observer if enabled
+        self._watcher = None
+
+    def build_node(self, name: str, seg: neo.Segment):
+
+        # The first source just produces filenames
+        return seg.make_source(name, self._generate_via_polling())
+
+    def _get_filename_queue(self) -> FiberQueue:
+        """
+        Returns an async queue with tuples of `([files], is_event)` where `is_event` indicates if this is a file changed
+        event (and we should wait for potentially more changes) or if these files were read on startup and should be
+        processed immediately.
+        """
+        q = FiberQueue(128)
+
+        if (self._watch_directory):
+
+            # Create a file watcher
+            self._watcher = Observer()
+            self._watcher.setDaemon(True)
+            self._watcher.setName("DirectoryWatcher")
+
+            event_handler = PatternMatchingEventHandler(patterns=[self._match_pattern])
+
+            def process_dir_change(event: FileSystemEvent):
+
+                # Push files into the queue indicating this is an event
+                q.put(([event.src_path], True))
+
+            event_handler.on_created = process_dir_change
+
+            self._watcher.schedule(event_handler, self._dir_to_watch, recursive=True)
+
+            self._watcher.start()
+
+        # Load the glob once and return
+        file_list = glob.glob(self._input_glob)
+        if self._sort_glob:
+            file_list = sorted(file_list)
+
+        if (self._max_files > 0):
+            file_list = file_list[:self._max_files]
+
+        logger.info("Found %d files in glob. Loading...", len(file_list))
+
+        # Push all to the queue and close it
+        q.put((file_list, False))
+
+        if (not self._watch_directory):
+            # Close the queue
+            q.close()
+
+        return q
+
+    def _generate_via_polling(self):
+
+        # Its a bit ugly, but utilize a filber queue to yield the thread. This will be improved in the future
+        file_queue = FiberQueue(128)
+
+        snapshot = EmptyDirectorySnapshot()
+
+        while (True):
+
+            # Get a new snapshot
+            new_snapshot = DirectorySnapshot(self._dir_to_watch, recursive=True)
+
+            # Take the diff from the last one
+            diff_events = DirectorySnapshotDiff(snapshot, new_snapshot)
+            snapshot = new_snapshot
+
+            files_to_process = filter_paths(diff_events.files_created, included_patterns=[self._input_glob])
+
+            if (self._sort_glob):
+                files_to_process = sorted(files_to_process)
+
+            # Convert from generator into list
+            files_to_process = list(files_to_process)
+
+            if (len(files_to_process) > 0):
+                # is_running = yield files_to_process
+                file_queue.put(files_to_process)
+
+            if (not self._watch_directory):
+                file_queue.close()
+
+            try:
+                batch_timeout = 5.0
```

Comment:
I would pull this out as a const at the top of the module or maybe make it a constructor argument